### ISSUE TITLE
[BUG] Invalid code emitted when array types are passed as template arguments to make_unique (etc)

### ISSUE THREAD

**Issue by bluetarpmedia** [hsutter/cppfront#197]
**Describe the bug**
Cppfront emits invalid code when an array type is passed as the template argument to functions like `std::make_unique` or the cpp2 `new` operator. Cppfront appears to parse the `T[]` template argument as a subscript operator call.

**To Reproduce**
Run cppfront with this pure cpp2 code:
```
main: () -> int = {

  buffer1 := std::make_unique<uint8_t[]>(100);
	
  buffer2 : std::unique_ptr<int[]> = std::make_unique<int[]>(200);
	
  buffer3 := std::make_shared<char[]>(3);
	
  buffer4 := new<int[]>(3);  // I assume this is valid cpp2 syntax
}
```

The emitted code is:
```
[[nodiscard]] auto main() -> int{

 auto buffer1 { std::make_unique<cpp2::assert_in_bounds(uint8_t, )>(100) }; 

 std::unique_ptr<cpp2::assert_in_bounds(int, )> buffer2 { std::make_unique<cpp2::assert_in_bounds(int, )>(200) }; 

 auto buffer3 { std::make_shared<cpp2::assert_in_bounds(char, )>(3) }; 

 auto buffer4 { cpp2_new<cpp2::assert_in_bounds(int, )>(3) }; 
}
```

**Expected Results**
The template arguments should remain as they were instead of being replaced with calls to `cpp2::assert_in_bounds`.

**Comment by hsutter** [comment]
Thanks for the well-written report!

tl;dr: Yes, this is intended behavior, but I should reject the code and give a better diagnostic.

Cpp2 does not support C-style arrays, and should use `array` instead. You can still use a `unique_ptr<T[]>` object created in Cpp1 code. Here's a repro I just tried:

```
#include <memory>
#include <array>

auto cpp1_func() {
    return std::make_unique<int[]>(10);
}

main: () -> int = {

    // ok, can consume a Cpp1 int[] specialization
    x := cpp1_func();

    // but in Cpp2 code, use std::array
    y := std::make_unique<std::array<int, 10>>();

}
```

However, I could emit a better diagnostic... I'm checking that in now, so that cases like the above will now print `error: subscript expression [ ] must not be empty (if you were trying to name a C-style array type, use 'std::array' instead) (at ']')`.
